torchdynamo/guards.py

```diff
@@ -404,11 +406,11 @@ def OBJECT_MUTATION(self, guard: Guard):
         mutation_guard.watch(self.get(guard.name), self.guarded_code)
 
     def GRAD_MODE(self, guard: Guard):
-        """Guard on the current value of torch.is_grad_enabled()"""
+        """Guard on the initial grad state"""
         assert guard.name == ""
         assert guard.source is GuardSource.GLOBAL
         code = None
-        if torch.is_grad_enabled():
```

Comment:
Can you explain this to me a little more? When would torch.is_grad_enabled() not line up with the state stored in `convert_frame.initial_grad_state` - is it because someone swapped grad state, but it happened in a skip / graph break? If so, shouldn't we actually be checking the current torch.is_grad_enabled()? Why have they deviated?

Reply:
Take a look at the unit test I added, if grad_enabled is True before the ```torch.set_grad_enabled``` and user's code sets it to ```False``` before graph break, then when this guard creating, the state is False which is not line up with the state stored in ```convert_frame.initial_grad_state``` which is True.

Reply:
But if user code says grad is false, why are we using true?

Reply:
We are guarding on the enter state, not the exit state, right?

Reply:
Yes, enter state. 

If the user code says grad is set to false, we should reflect that in the guard.

What I don't understand is how torch.is_grad_enabled() can give us the wrong value for the state of grad... Why do we want the initial value, instead of the user set one?